### PR TITLE
Masonry: delay updating React scroll state until scroll events stop coming in

### PR DESCRIPTION
**Pull request by valbooth28**
### Summary
**TLDR:** This changes the `updateScrollPosition` strategy in Masonry from: making sure that updating isn't called multiple times w/in 100ms, to: not calling `updateScrollPosition` at all until scroll callbacks stop occurring.

The browser API for scroll callbacks is called multiple times per scroll events. Therefore, any code called within the scroll handler callback can very easily slow down scroll. This seems to be the idea behind the `throttle`ing here, to ensure multiple callbacks aren't called within 100ms of each other: 

https://github.com/pinterest/gestalt/blob/6021664f7341a7f3f7a52887e388c1203da77e41/packages/gestalt/src/Masonry.js#L135

https://github.com/valbooth28/gestalt/blob/aefb6127eacbe19a23bdfdc0ecf8d326c4eb6dac/packages/gestalt/src/throttle.js#L10

At the start of a scroll, however, no scroll callback has been called, so the React state update is called _immediately_, slowing down the 1,2,3... invocations of the `addEventListener('scroll'` events. When scrolling a page during Performance Profiling in Chrome DevTools, it is explicitly highlighted as slowing down the handler:

Scroll Before
Scroll Before 2

### Links

- [Jira](https://jira.pinadmin.com/browse/PERF-10133)
- Further documentation about Scroll RUM data logging seen in the gifs available upon request

#### Before
_See above screenshots as well. PWT values are milliseconds per frame (or 1/FPS)_

scroll PWT before

#### After
_PWT values are milliseconds per frame (or 1/FPS)_
Scroll After Code

scroll PWT after

### Checklist

- [ ] Added unit and Flow Tests
- [ ] Added documentation + accessibility tests
- [ ] Verified accessibility: keyboard & screen reader interaction
- [ ] Checked dark mode, responsiveness, and right-to-left support
- [ ] Checked stakeholder feedback (e.g. Gestalt designers)

### CODE REVIEW

File: packages/gestalt/src/Masonry.js [hunk 1]
```diff
@@ -145,7 +142,24 @@ export default class Masonry<T: { ... }> extends ReactComponent<Props<T>, State<
...
     });
-  });
+  };
+
+  scrollTimeoutFn: ? TimeoutID = null;
+
+  // The browser scroll event API will fire LOTS of events. This waits until the events
+  // stop coming in to set the final scroll position
+  // $FlowFixMe[signature-verification-failure]
+  handleScroll = () => {
```
Comment (valbooth28) [inline packages/gestalt/src/Masonry.js:152]: I tried to have a void function signature here:

File: packages/gestalt/src/Masonry.js [hunk 2]
```diff
@@ -145,7 +142,24 @@ export default class Masonry<T: { ... }> extends ReactComponent<Props<T>, State<
...
+    if (this.scrollTimeoutFn) {
+      clearTimeout(this.scrollTimeoutFn);
+    }
+    this.scrollTimeoutFn = setTimeout(this.updateScrollPosition, 100);
+  };
+
+  // $FlowFixMe[signature-verification-failure]
+  clearScrollTimeout = (): void => {
```
Comment (valbooth28) [inline packages/gestalt/src/Masonry.js:160]: Likewise: 
Screen Shot 2021-11-09 at 2 34 05 PM

File: packages/gestalt/src/Masonry.js [hunk 3]
```diff
@@ -145,7 +142,24 @@ export default class Masonry<T: { ... }> extends ReactComponent<Props<T>, State<
...
+  scrollTimeoutFn: ? TimeoutID = null;
+
+  // The browser scroll event API will fire LOTS of events. This waits until the events
+  // stop coming in to set the final scroll position
+  // $FlowFixMe[signature-verification-failure]
+  handleScroll = () => {
+    if (this.scrollTimeoutFn) {
+      clearTimeout(this.scrollTimeoutFn);
+    }
+    this.scrollTimeoutFn = setTimeout(this.updateScrollPosition, 100);
```
Comment (valbooth28) [inline packages/gestalt/src/Masonry.js:156]: 100ms based on previous throttle value, happy to change.

File: packages/gestalt/src/Masonry.js (hunk 1 above)
Comment (dangerismycat) [reply packages/gestalt/src/Masonry.js:152]: Hm…that's odd. I'm not super familiar with typing class methods, but that looks like it should be the correct syntax to me. You could try more explicit syntax, which may work:
```js
handleScroll: () => void = () => {
  ...
}
```
TBH I don't like this syntax as I find it difficult to read, but if it works… ¯\_(ツ)_/¯

File: packages/gestalt/src/Masonry.js (hunk 3 above)
Comment (dangerismycat) [reply packages/gestalt/src/Masonry.js:156]: I'm not sure if it should change, but do you mind making it a top-of-file const with a descriptive name? If that is a knob we'll potentially want to tweak in the future, it'd be nice to make it easier to find.

File: packages/gestalt/src/Masonry.js [hunk 4]
```diff
@@ -145,7 +142,24 @@ export default class Masonry<T: { ... }> extends ReactComponent<Props<T>, State<
     this.setState({
       scrollTop: getScrollPos(scrollContainer),
     });
-  });
+  };
+
+  scrollTimeoutFn: ?TimeoutID = null;
```
Comment (dangerismycat) [inline packages/gestalt/src/Masonry.js:147]: nit: this isn't a function, it's a timeout id, so the name seems a bit misleading

**Review by dangerismycat — COMMENTED** [review]
This seems fine to me. @liuyenwei do you have any concerns?

File: packages/gestalt/src/Masonry.js (hunk 1 above)
Comment (valbooth28) [reply packages/gestalt/src/Masonry.js]: The double parens did it - thanks!

**Comment by chrislloyd** [comment]
This is a large enough change that it should be run as an experiment to validate user behavior. Synthetic dev-approximated results are promising, but lets see what happens in the wild. Happy to approve if you put this behind a flag.

**Review by chrislloyd — APPROVED** [review]
Thanks for the speedy turnaround! LGTM

File: packages/gestalt/src/Masonry.js [hunk 5]
```diff
@@ -87,6 +87,7 @@ type Props<T> = {|
    * This is required if the grid is expected to be scrollable.
    */
   scrollContainer?: () => HTMLElement,
+  deferScrollPositionUpdate?: boolean,
```
Comment (dangerismycat) [inline packages/gestalt/src/Masonry.js:90]: A couple of small requests:
- Can you add a `_` prefix to the prop name? We're in the process of converting our docs props tables to auto-generate from the component Props types, and we're likely to implement screening out undocumented props by that prefix.
- Can you add an explanatory comment with context around this prop?

**Review by dangerismycat — APPROVED** [review]

File: packages/gestalt/src/Masonry.js (hunk 5 above)
Comment (valbooth28) [reply packages/gestalt/src/Masonry.js]: Friendly callout that you'll need to disable the dangling eslint rule when you do that: 
Screen Shot 2021-11-16 at 4 21 08 PM
